README.md
# PythaiNAV: ทำให้การดึงข้อมูลกองทุนไทยเป็นเรื่องง่าย
<!-- ALL-CONTRIBUTORS-BADGE:START - Do not remove or modify this section -->
[![All Contributors](https://img.shields.io/badge/all_contributors-3-orange.svg?style=flat-square)](#contributors-)
<!-- ALL-CONTRIBUTORS-BADGE:END -->
![Tests](https://github.com/CircleOnCircles/pythainav/workflows/Tests/badge.svg?branch=master)
[![codecov](https://codecov.io/gh/CircleOnCircles/pythainav/branch/develop/graph/badge.svg)](https://codecov.io/gh/CircleOnCircles/pythainav)
[![Codacy Badge](https://api.codacy.com/project/badge/Grade/f868488db4ba4266a112c3432301c6b4)](https://www.codacy.com/manual/nutchanon/pythainav?utm_source=github.com&amp;utm_medium=referral&amp;utm_content=CircleOnCircles/pythainav&amp;utm_campaign=Badge_Grade)



![cover image](https://github.com/CircleOnCircles/pythainav/raw/master/extra/pythainav.png)



> อยากชวนทุกคนมาร่วมพัฒนา ติชม แนะนำ เพื่อให้ทุกคนเข้าถึงข้อมูลการง่ายขึ้น [เริ่มต้นได้ที่นี้](https://github.com/CircleOnCircles/pythainav/issues) หรือเข้ามา Chat ใน [Discord](https://discord.gg/jjuMcKZ) ได้เลย 😊

📖 Documentation is here. คู่มือการใช้งานอยู่ที่นี่ <https://pythainav.nutchanon.org/>

## Get Started - เริ่มต้นใช้งาน

```bash
$ pip install pythainav
```

```python
import pythainav as nav

nav.get("KT-PRECIOUS")
> Nav(value=4.2696, updated='20/01/2020', tags={'latest'}, fund='KT-PRECIOUS')

nav.get("TISTECH-A", date="1 week ago")
> Nav(value=12.9976, updated='14/01/2020', tags={}, fund='TISTECH-A')

nav.get_all("TISTECH-A")
> [Nav(value=12.9976, updated='21/01/2020', tags={}, fund='TISTECH-A'), Nav(value=12.9002, updated='20/01/2020', tags={}, fund='TISTECH-A'), ...]

nav.get_all("KT-PRECIOUS", asDataFrame=True)
> pd.DataFrame [2121 rows x 4 columns]
```

## Source of Data - ที่มาข้อมูล

ดูจาก <https://pythainav.nutchanon.org/datasource/>

## Disclaimer

เราไม่รับประกันความเสียหายใดๆทั้งสิ้นที่เกิดจาก แหล่งข้อมูล, library, source code,sample code, documentation, library dependencies และอื่นๆ

## Contributors ✨

Thanks goes to these wonderful people ([emoji key](https://allcontributors.org/docs/en/emoji-key)):

<!-- ALL-CONTRIBUTORS-LIST:START - Do not remove or modify this section -->
<!-- prettier-ignore-start -->
<!-- markdownlint-disable -->
<table>
  <tr>
    <td align="center"><a href="http://nutchanon.org"><img src="https://avatars2.githubusercontent.com/u/8089231?v=4" width="100px;" alt=""/><br /><sub><b>Nutchanon Ninyawee</b></sub></a><br /><a href="https://github.com/CircleOnCircles/pythainav/commits?author=CircleOnCircles" title="Code">💻</a> <a href="#infra-CircleOnCircles" title="Infrastructure (Hosting, Build-Tools, etc)">🚇</a></td>
    <td align="center"><a href="https://github.com/sctnightcore"><img src="https://avatars2.githubusercontent.com/u/23263315?v=4" width="100px;" alt=""/><br /><sub><b>sctnightcore</b></sub></a><br /><a href="https://github.com/CircleOnCircles/pythainav/commits?author=sctnightcore" title="Code">💻</a> <a href="#talk-sctnightcore" title="Talks">📢</a> <a href="#ideas-sctnightcore" title="Ideas, Planning, & Feedback">🤔</a></td>
    <td align="center"><a href="https://github.com/angonyfox"><img src="https://avatars3.githubusercontent.com/u/1295513?v=4" width="100px;" alt=""/><br /><sub><b>angonyfox</b></sub></a><br /><a href="https://github.com/CircleOnCircles/pythainav/commits?author=angonyfox" title="Code">💻</a> <a href="https://github.com/CircleOnCircles/pythainav/commits?author=angonyfox" title="Tests">⚠️</a></td>
  </tr>
</table>

<!-- markdownlint-enable -->
<!-- prettier-ignore-end -->
<!-- ALL-CONTRIBUTORS-LIST:END -->

This project follows the [all-contributors](https://github.com/all-contributors/all-contributors) specification. Contributions of any kind welcome!

## Related and Alternative

ตัวที่ใช้ได้ในปัจจุบัน 22/02/20
* [uncleEngineer](https://github.com/UncleEngineer/uncleengineer) - ดึงข้อมูลหุ้น ณ ปัจจุบัน
* [pandas-datareader](https://www.patanasongsivilai.com/blog/stock-thai-python/) - ดึงข้อมูลหุ้นย้อนหลังผ่าน `pdr.get_data_yahoo`


CHANGELOG.md
# Release Note

## Pending
### Add
 - new data source Sec.or.th

## 0.1.5 - 9 March 2020

### Fixes
 - down pandas min version to 0.25.3 for google colab support

## 0.1.5 - 9 March 2020

### Added
 - get() now support date parameter
 - add get_all()
 - update docs, readme, source of data

## 0.1.3 - 25 Jan 2020

### Fixes
 - dataclasses error on python 3.6

## 0.1.2 - 25 Jan 2020

### Features

 - Change to api call = more lightweight + faster

### Fixes
 - Add compatibility to python 3.6

## 0.1.1 - 24 Jan 2020

### Misc.
 - Setup skeleton for future development
 - Refactor the code


tests/test_core.py
import pythainav as nav


def test_get_nav():
    kt_nav = nav.get("KT-PRECIOUS")
    print(kt_nav)
    assert kt_nav.value >= 0


def test_get_nav_with_date():
    kt_nav = nav.get("KT-PRECIOUS", date="1 week ago")
    print(kt_nav)
    assert kt_nav.value >= 0


def test_get_all():
    kt_navs = nav.get_all("KT-PRECIOUS")
    print(kt_navs)
    assert len(kt_navs) >= 0


def test_get_all_pandas():
    import pandas as pd

    df = nav.get_all("KT-PRECIOUS", asDataFrame=True)
    assert isinstance(df, pd.DataFrame)


def test_sec_source():
    from decouple import config

    subs_key = {"fundfactsheet": config("FUND_FACTSHEET_KEY"), "funddailyinfo": config("FUND_DAILY_INFO_KEY")}

    # should auto convert to friday if it is a weekend
    kt_nav = nav.get("KT-PRECIOUS", source="sec", subscription_key=subs_key)
    print(kt_nav)
    if kt_nav is not None:
        assert kt_nav.value >= 0

    kt_nav = nav.get("KT-PRECIOUS", source="sec", subscription_key=subs_key, date="03/04/2020")
    print(kt_nav)
    assert kt_nav.value >= 0


tests/sources/conftest.py
import json
import re

import pytest

from furl import furl

import httpretty

from tests.factories.dailynav import AMCInfoFactory
from tests.factories.dailynav import DailyNavFactory
from tests.factories.search_class_fund import SearchClassFundFactory
from tests.factories.search_fund import SearchFundFactory


@pytest.fixture
def subscription_key():
    subscription_key = {"fundfactsheet": "fact_key", "funddailyinfo": "daily_key"}
    return subscription_key


@pytest.fixture
def dataset():
    dataset = {}
    dataset["search_fund_data"] = SearchFundFactory.build_batch(5)
    dataset["search_class_fund_data"] = SearchClassFundFactory.build_batch(5)
    dataset["dailynav_data"] = DailyNavFactory.create()

    multi_class_dailynav_data = DailyNavFactory.create(class_fund=True)
    multi_class_amc_info = AMCInfoFactory(class_fund=True)
    multi_class_dailynav_data["amc_info"] = [multi_class_amc_info]
    dataset["multi_class_dailynav_data"] = multi_class_dailynav_data
    return dataset


@pytest.fixture(autouse=True)
def set_global_test_data(request):
    httpretty.reset()
    if not httpretty.is_enabled():
        httpretty.enable()

    dataset = request.getfixturevalue("dataset")

    #
    # Sec
    #
    # FundFactsheet 3
    httpretty.register_uri(
        httpretty.POST, "https://api.sec.or.th/FundFactsheet/fund", body=json.dumps(dataset["search_fund_data"])
    )

    # FundFactsheet 21
    httpretty.register_uri(
        httpretty.POST,
        "https://api.sec.or.th/FundFactsheet/fund/class_fund",
        body=json.dumps(dataset["search_class_fund_data"]),
    )

    # FundDailyInfo 1
    httpretty.register_uri(
        httpretty.GET,
        re.compile("https://api.sec.or.th/FundDailyInfo/.*/dailynav/.*"),
        body=json.dumps(dataset["dailynav_data"]),
    )

    #
    # Onde
    #
    base_url = {
        "fundfactsheet": furl(
            "http://dataexchange.onde.go.th/api/ApiProxy/Data/3C154331-4622-406E-94FB-443199D35523/f2529128-0332-44a0-9066-034093b07837/fund"
        )
    }
    # FundFactsheet 3
    httpretty.register_uri(httpretty.POST, base_url["fundfactsheet"].url, body=json.dumps(dataset["search_fund_data"]))

    # FundFactsheet 21
    httpretty.register_uri(
        httpretty.POST,
        base_url["fundfactsheet"].copy().add(path=["class_fund"]).url,
        body=json.dumps(dataset["search_class_fund_data"]),
    )

    # FundDailyInfo 1
    httpretty.register_uri(
        httpretty.GET,
        re.compile(
            "http://dataexchange.onde.go.th/api/ApiProxy/Data/92b67f7e-023e-4ce8-b4ba-08989d44ff78/ed6867e7-2d97-49e3-b25b-8bb0edb18e1c/.*/dailynav/.*"
        ),
        body=json.dumps(dataset["dailynav_data"]),
    )
    yield
    httpretty.disable()


tests/sources/test_sec.py
import datetime
import json
import re

from unittest.mock import patch

import dateparser
import pytest
import requests

import httpretty

from pythainav.nav import Nav
from pythainav.sources import Sec


def test_no_subscription_key():
    with pytest.raises(ValueError):
        Sec()


def test_subscription_key_is_none():
    subscription_key = None
    with pytest.raises(ValueError):
        Sec(subscription_key)

    with pytest.raises(ValueError):
        Sec(subscription_key=subscription_key)


def test_subscription_key_missing_key():

    with pytest.raises(ValueError):
        Sec(subscription_key={"wrong_name": "some_key"})

    with pytest.raises(ValueError):
        Sec(subscription_key={"fundfactsheet": "some_key"})

    with pytest.raises(ValueError):
        Sec(subscription_key={"funddailyinfo": "some_key"})

    test_sec = Sec(subscription_key={"fundfactsheet": "some_key", "funddailyinfo": "some_key"})
    assert list(test_sec.subscription_key.keys()) == ["fundfactsheet", "funddailyinfo"]


#
#   search_fund
#
def test_search_fund_setting_headers(subscription_key):
    source = Sec(subscription_key=subscription_key)
    source.search_fund("FUND")

    # contain Ocp-Apim-Subscription-Key in header
    assert "Ocp-Apim-Subscription-Key" in httpretty.last_request().headers
    assert httpretty.last_request().headers["Ocp-Apim-Subscription-Key"] == subscription_key["fundfactsheet"]


def test_search_fund_invalid_key(subscription_key):
    httpretty.reset()
    error_responses = [
        httpretty.Response(
            status=401,
            body=json.dumps(
                {
                    "statusCode": 401,
                    "message": "Access denied due to invalid subscription key. Make sure to provide a valid key for an active subscription.",
                }
            ),
        )
    ]

    httpretty.register_uri(httpretty.POST, "https://api.sec.or.th/FundFactsheet/fund", responses=error_responses)
    source = Sec(subscription_key=subscription_key)
    with pytest.raises(requests.exceptions.HTTPError):
        source.search_fund("FUND")


def test_search_fund_success_with_content(subscription_key, dataset):
    source = Sec(subscription_key=subscription_key)
    result = source.search_fund("FUND")

    assert result == dataset["search_fund_data"]


def test_search_fund_no_content(subscription_key):
    # status code 204
    httpretty.reset()

    httpretty.register_uri(httpretty.POST, "https://api.sec.or.th/FundFactsheet/fund", status=204)

    source = Sec(subscription_key=subscription_key)
    result = source.search_fund("FUND")
    assert result is None


#
#   search_class_fund
#
def test_search_class_fund_setting_headers(subscription_key):
    source = Sec(subscription_key=subscription_key)
    source.search_class_fund("FUND")

    # contain Ocp-Apim-Subscription-Key in header
    assert "Ocp-Apim-Subscription-Key" in httpretty.last_request().headers
    assert httpretty.last_request().headers["Ocp-Apim-Subscription-Key"] == subscription_key["fundfactsheet"]


def test_search_class_fund_invalid_key(subscription_key):
    httpretty.reset()
    error_responses = [
        httpretty.Response(
            status=401,
            body=json.dumps(
                {
                    "statusCode": 401,
                    "message": "Access denied due to invalid subscription key. Make sure to provide a valid key for an active subscription.",
                }
            ),
        )
    ]

    httpretty.register_uri(httpretty.POST, "https://api.sec.or.th/FundFactsheet/fund/class_fund", responses=error_responses)
    source = Sec(subscription_key=subscription_key)
    with pytest.raises(requests.exceptions.HTTPError):
        source.search_class_fund("FUND")


def test_search_class_fund_success_with_content(subscription_key, dataset):
    source = Sec(subscription_key=subscription_key)
    result = source.search_class_fund("FUND")

    assert result == dataset["search_class_fund_data"]


def test_search_class_fund_no_content(subscription_key):
    # status code 204
    httpretty.reset()

    httpretty.register_uri(httpretty.POST, "https://api.sec.or.th/FundFactsheet/fund/class_fund", status=204)

    source = Sec(subscription_key=subscription_key)
    result = source.search_class_fund("FUND")
    assert result is None


#
#   search
#
@patch("pythainav.sources.Sec.search_class_fund")
@patch("pythainav.sources.Sec.search_fund")
def test_search_result(mock_search_fund, mock_search_class_fund, subscription_key, dataset):
    # search_fund found fund
    mock_search_fund.return_value = dataset["search_fund_data"]
    source = Sec(subscription_key=subscription_key)
    result = source.search("FUND")

    assert result == dataset["search_fund_data"]

    # search_fund return empty
    mock_search_fund.return_value = None
    mock_search_class_fund.return_value = dataset["search_class_fund_data"]
    result = source.search("FUND")
    assert mock_search_class_fund.called
    assert result == dataset["search_class_fund_data"]

    # both return empty
    mock_search_fund.return_value = None
    mock_search_class_fund.return_value = None
    result = source.search("FUND")
    assert result is None


#
#   list
#
def test_list_result(subscription_key, dataset):
    source = Sec(subscription_key=subscription_key)
    result = source.list()

    assert len(result) == len(dataset["search_fund_data"])


#
#   get_nav_from_fund_id
#
def test_get_nav_from_fund_id_success_with_content(subscription_key, dataset):
    # status code 200
    expect_return = Nav(
        value=float(dataset["dailynav_data"]["last_val"]),
        updated=dateparser.parse(dataset["dailynav_data"]["nav_date"]),
        tags={},
        fund="FUND_ID",
    )

    nav_date = datetime.date(2020, 1, 1)
    source = Sec(subscription_key=subscription_key)
    result = source.get_nav_from_fund_id("FUND_ID", nav_date)

    assert result == expect_return


def test_get_nav_from_fund_id_no_content(subscription_key):
    # status code 204
    httpretty.reset()

    httpretty.register_uri(httpretty.GET, re.compile("https://api.sec.or.th/FundDailyInfo/.*/dailynav/.*"), status=204)
    nav_date = datetime.date(2020, 1, 1)
    source = Sec(subscription_key=subscription_key)
    result = source.get_nav_from_fund_id("FUND_ID", nav_date)
    assert result is None


def test_get_nav_from_fund_id_multi_class(subscription_key, dataset):
    httpretty.reset()

    httpretty.register_uri(
        httpretty.GET,
        re.compile("https://api.sec.or.th/FundDailyInfo/.*/dailynav/.*"),
        body=json.dumps(dataset["multi_class_dailynav_data"]),
    )

    fund_name = "FUND_ID"
    remark_en = dataset["multi_class_dailynav_data"]["amc_info"][0]["remark_en"]
    multi_class_nav = {k.strip(): float(v) for x in remark_en.split("/") for k, v in [x.split("=")]}
    expect_return = []
    for fund_name, nav_val in multi_class_nav.items():
        n = Nav(
            value=float(nav_val),
            updated=dateparser.parse(dataset["multi_class_dailynav_data"]["nav_date"]),
            tags={},
            fund=fund_name,
        )
        n.amount = dataset["multi_class_dailynav_data"]["net_asset"]
        expect_return.append(n)

    nav_date = datetime.date(2020, 1, 1)
    source = Sec(subscription_key=subscription_key)
    result = source.get_nav_from_fund_id(fund_name, nav_date)
    assert result == expect_return


#
#   get
#
def test_get_params(subscription_key):
    nav_date = datetime.date(2020, 1, 1)
    source = Sec(subscription_key=subscription_key)

    # date: str
    try:
        source.get("FUND", nav_date.isoformat())
    except Exception:
        pytest.fail("raise exception unexpectedly")

    # Empty Fund
    with pytest.raises(ValueError):
        source.get("", nav_date.isoformat())

    # date: datetime.date
    try:
        source.get("FUND", nav_date)
    except Exception:
        pytest.fail("raise exception unexpectedly")

    # date: datetime.datetime
    try:
        source.get("FUND", datetime.datetime.combine(nav_date, datetime.datetime.min.time()))
    except Exception:
        pytest.fail("raise exception unexpectedly")


pythainav/sources.py
import base64
import datetime

from abc import ABC
from abc import abstractmethod
from functools import lru_cache
from typing import List

import dateparser
import requests

from furl import furl

from .nav import Nav
from .utils.date import convert_buddhist_to_gregorian
from .utils.date import date_range


class Source(ABC):
    @abstractmethod
    def get(self, fund: str):
        pass

    @abstractmethod
    def list(self):
        pass


class Finnomena(Source):
    base = furl("https://www.finnomena.com/fn3/api/fund/")

    def __init__(self):
        super().__init__()

    def _find_earliest(self, navs: List[Nav], date: str):
        navs.sort(key=lambda x: x.updated, reverse=True)

        date = dateparser.parse(date)
        for nav in navs:
            if date >= nav.updated:
                return nav
        else:
            return None

    def get(self, fund: str, date: str = None):

        if date:
            navs = self.get_range(fund)
            return self._find_earliest(navs, date)

        name2fund = self.list()

        url = self.base / "nav" / "latest"
        url.args["fund"] = name2fund[fund]["id"]

        # convert to str
        url = url.url

        nav = requests.get(url).json()
        nav = Nav(value=float(nav["value"]), updated=dateparser.parse(nav["nav_date"]), tags={"latest"}, fund=fund)
        return nav

    # cache here should be sensible since the fund is not regulary update
    # TODO: change or ttl cache with timeout = [1 hour, 1 day]
    @lru_cache(maxsize=1024)
    def get_range(self, fund: str, period="SI"):
        name2fund = self.list()

        url = self.base / "nav" / "q"
        url.args["fund"] = name2fund[fund]["id"]
        url.args["range"] = period

        # convert to str
        url = url.url

        navs_response = requests.get(url).json()

        navs = []
        for nav_resp in navs_response:
            nav = Nav(value=float(nav_resp["value"]), updated=dateparser.parse(nav_resp["nav_date"]), tags={}, fund=fund)
            nav.amount = nav_resp["amount"]
            navs.append(nav)

        return navs

    # cache here should be sensible since the fund is not regulary update
    # TODO: change or ttl cache with timeout = [1 hour, 1 day]
    @lru_cache(maxsize=1)
    def list(self):
        url = self.base / "public" / "list"
        url = url.url
        funds = requests.get(url).json()
        return {fund["short_code"]: fund for fund in funds}

    # def _list(self, )


class Sec(Source):
    base = furl("https://api.sec.or.th/")

    def __init__(self, subscription_key: dict = None):
        super().__init__()
        if subscription_key is None:
            # TODO: Create specific exception for this
            raise ValueError("Missing subscription key")
        if not all([True if key in subscription_key else False for key in ["fundfactsheet", "funddailyinfo"]]):
            raise ValueError("subscription_key must contain 'fundfactsheet' and 'funddailyinfo' key")
        self.subscription_key = subscription_key
        self.headers = {
            "Content-Type": "application/json",
        }
        self.base_url = {
            "fundfactsheet": self.base.copy().add(path=["FundFactsheet", "fund"]),
            "funddailyinfo": self.base.copy().add(path="FundDailyInfo"),
        }
        self.session = requests.Session()
        self.session.headers.update(self.headers)

    def __get_api_data(self, url, headers=None, subscription_key="fundfactsheet"):
        if headers is None:
            headers = self.headers
        headers.update({"Ocp-Apim-Subscription-Key": self.subscription_key[subscription_key]})
        response = self.session.get(url, headers=headers)
        # check status code
        response.raise_for_status()
        if response.status_code == 200:
            if response.headers["content-length"] == "0":
                raise requests.exceptions.ConnectionError("No data received")
            return response.json()
        # No content
        elif response.status_code == 204:
            return None

    def get(self, fund: str, date: str = None):
        if date:
            if isinstance(date, str):
                query_date = dateparser.parse(date).date()
            elif isinstance(date, datetime.date):
                query_date = date
            elif isinstance(date, datetime.datetime):
                query_date = date.date()
        else:
            # TODO: Upgrade to smarter https://stackoverflow.com/questions/2224742/most-recent-previous-business-day-in-python
            # PS. should i add pandas as dep? it's so largeee.
            query_date = last_bus_day = datetime.date.today()
            wk_day = datetime.date.weekday(last_bus_day)
            if wk_day > 4:  # if it's Saturday or Sunday
                last_bus_day = last_bus_day - datetime.timedelta(days=wk_day - 4)  # then make it Friday
            query_date = last_bus_day

        if not fund:
            raise ValueError("Must specify fund")

        list_fund = self.search(fund)
        if list_fund:
            fund_info = list_fund.pop(0)
            fund_id = fund_info["proj_id"]
            nav = self.get_nav_from_fund_id(fund_id, query_date)

            if isinstance(nav, Nav):
                nav.fund = fund_info["proj_abbr_name"]
                if query_date == datetime.date.today():
                    nav.tags = {"latest"}
                return nav
            else:
                return nav
        else:
            # Fund not found
            # due to query_date is a week day that also a holiday
            return None

    def get_range(self, fund: str, period="SI"):
        list_fund = self.search(fund)
        if list_fund:
            fund_info = list_fund.pop(0)
            today = datetime.date.today()
            if period == "SI":
                if fund_info["regis_date"] != "-":
                    inception_date = dateparser.parse(fund_info["regis_date"]).date()
                    data_date = date_range(inception_date, today)
                else:
                    date_date = [today]
            else:
                query_date = dateparser.parse(period).date()
                data_date = date_range(query_date, today)
            list_nav = []
            # Remove weekend
            data_date = [dd for dd in date_date if dd.isoweekday() not in [6, 7]]
            for dd in data_date:
                nav = self.get_nav_from_fund_id(fund, dd)
                if nav:
                    list_nav.append(nav)
            return list_nav
        else:
            # Fund not found
            return None

    @lru_cache(maxsize=1024)
    def get_nav_from_fund_id(self, fund_id: str, nav_date: datetime.date):
        url = self.base_url["funddailyinfo"].copy().add(path=[fund_id, "dailynav", nav_date.isoformat()]).url
        headers = self.headers
        headers.update({"Ocp-Apim-Subscription-Key": self.subscription_key["funddailyinfo"]})
        response = self.session.get(url, headers=headers)
        # check status code
        response.raise_for_status()
        if response.status_code == 200:
            if response.headers["content-length"] == "0":
                raise requests.exceptions.ConnectionError("No data received")
            result = response.json()
            # Multi class fund
            if float(result["last_val"]) == 0.0 and float(result["previous_val"]) == 0:
                remark_en = result["amc_info"][0]["remark_en"]
                multi_class_nav = {k.strip(): float(v) for x in remark_en.split("/") for k, v in [x.split("=")]}
                list_nav = []
                for fund_name, nav_val in multi_class_nav.items():
                    n = Nav(value=float(nav_val), updated=dateparser.parse(result["nav_date"]), tags={}, fund=fund_name)
                    n.amount = result["net_asset"]
                    list_nav.append(n)
                return list_nav
            else:
                n = Nav(value=float(result["last_val"]), updated=dateparser.parse(result["nav_date"]), tags={}, fund=fund_id)
                n.amount = result["net_asset"]
                return n
        # No content
        elif response.status_code == 204:
            return None

    @lru_cache(maxsize=1024)
    def list(self):
        return self.search_fund(name="")

    def search(self, name: str):
        result = self.search_fund(name)
        if result is None:
            result = self.search_class_fund(name)
        return result

    @lru_cache(maxsize=1024)
    def search_fund(self, name: str):
        url = self.base_url["fundfactsheet"].url
        headers = self.headers
        headers.update({"Ocp-Apim-Subscription-Key": self.subscription_key["fundfactsheet"]})
        response = self.session.post(url, headers=headers, json={"name": name})
        # check status code
        response.raise_for_status()
        if response.status_code == 200:
            if response.headers["content-length"] == "0":
                raise requests.exceptions.ConnectionError("No data received")
            return response.json()
        # No content
        elif response.status_code == 204:
            return None

    @lru_cache(maxsize=1024)
    def search_class_fund(self, name: str):
        url = self.base_url["fundfactsheet"].copy().add(path="class_fund").url
        headers = self.headers
        headers.update({"Ocp-Apim-Subscription-Key": self.subscription_key["fundfactsheet"]})
        response = self.session.post(url, headers=headers, json={"name": name})
        # check status code
        response.raise_for_status()
        if response.status_code == 200:
            if response.headers["content-length"] == "0":
                raise requests.exceptions.ConnectionError("No data received")
            return response.json()
        # No content
        elif response.status_code == 204:
            return None

    def list_amc(self):
        url = self.base_url["fundfactsheet"].copy().add(path="amc").url
        result = self.__get_api_data(url)
        return result

    def list_fund_under_amc(self, amc_id):
        if amc_id is None and not amc_id:
            raise ValueError("Missing amc_id")
        url = self.base_url["fundfactsheet"].copy().add(path=["amc", amc_id]).url
        result = self.__get_api_data(url)
        return result

    def get_fund_factsheet_url(self, fund_id):
        if not fund_id:
            raise ValueError("Must specify fund")
        url = self.base_url["fundfactsheet"].copy().add(path=[fund_id, "URLs"]).url
        result = self.__get_api_data(url)
        return result

    def get_fund_ipo(self, fund_id):
        if not fund_id:
            raise ValueError("Must specify fund")
        url = self.base_url["fundfactsheet"].copy().add(path=[fund_id, "IPO"]).url
        result = self.__get_api_data(url)
        return result

    def get_fund_investment(self, fund_id):
        if not fund_id:
            raise ValueError("Must specify fund")
        url = self.base_url["fundfactsheet"].copy().add(path=[fund_id, "investment"]).url
        result = self.__get_api_data(url)
        return result

    def get_fund_project_type(self, fund_id):
        if not fund_id:
            raise ValueError("Must specify fund")
        url = self.base_url["fundfactsheet"].copy().add(path=[fund_id, "project_type"]).url
        result = self.__get_api_data(url)
        return result

    def get_fund_policy(self, fund_id):
        if not fund_id:
            raise ValueError("Must specify fund")
        url = self.base_url["fundfactsheet"].copy().add(path=[fund_id, "policy"]).url
        result = self.__get_api_data(url)
        if "investment_policy_desc" in result and len(result["investment_policy_desc"]):
            result["investment_policy_desc"] = base64.b64decode(result["investment_policy_desc"]).decode("utf-8")
        return result

    def get_fund_specification(self, fund_id):
        if not fund_id:
            raise ValueError("Must specify fund")
        url = self.base_url["fundfactsheet"].copy().add(path=[fund_id, "specification"]).url
        result = self.__get_api_data(url)
        return result

    def get_fund_feeder_fund(self, fund_id):
        if not fund_id:
            raise ValueError("Must specify fund")
        url = self.base_url["fundfactsheet"].copy().add(path=[fund_id, "feeder_fund"]).url
        result = self.__get_api_data(url)
        return result

    def get_fund_redemption(self, fund_id):
        if not fund_id:
            raise ValueError("Must specify fund")
        url = self.base_url["fundfactsheet"].copy().add(path=[fund_id, "redemption"]).url
        result = self.__get_api_data(url)
        return result

    def get_fund_suitability(self, fund_id):
        if not fund_id:
            raise ValueError("Must specify fund")
        url = self.base_url["fundfactsheet"].copy().add(path=[fund_id, "suitability"]).url
        result = self.__get_api_data(url)
        for key in ["fund_suitable_desc", "fund_not_suitable_desc", "important_notice", "risk_spectrum_desc"]:
            if key in result:
                result[key] = base64.b64decode(result[key]).decode("utf-8")
        return result

    def get_fund_risk(self, fund_id):
        if not fund_id:
            raise ValueError("Must specify fund")
        url = self.base_url["fundfactsheet"].copy().add(path=[fund_id, "risk"]).url
        result = self.__get_api_data(url)
        return result

    def get_fund_asset(self, fund_id):
        if not fund_id:
            raise ValueError("Must specify fund")
        url = self.base_url["fundfactsheet"].copy().add(path=[fund_id, "asset"]).url
        result = self.__get_api_data(url)
        return result

    def get_fund_turnover_ratio(self, fund_id):
        if not fund_id:
            raise ValueError("Must specify fund")
        url = self.base_url["fundfactsheet"].copy().add(path=[fund_id, "turnover_ratio"]).url
        result = self.__get_api_data(url)
        return result

    def get_fund_return(self, fund_id):
        if not fund_id:
            raise ValueError("Must specify fund")
        url = self.base_url["fundfactsheet"].copy().add(path=[fund_id, "return"]).url
        result = self.__get_api_data(url)
        return result

    def get_fund_buy_and_hold(self, fund_id):
        if not fund_id:
            raise ValueError("Must specify fund")
        url = self.base_url["fundfactsheet"].copy().add(path=[fund_id, "buy_and_hold"]).url
        result = self.__get_api_data(url)
        return result

    def get_fund_benchmark(self, fund_id):
        if not fund_id:
            raise ValueError("Must specify fund")
        url = self.base_url["fundfactsheet"].copy().add(path=[fund_id, "benchmark"]).url
        result = self.__get_api_data(url)
        return result

    def get_fund_compare(self, fund_id):
        if not fund_id:
            raise ValueError("Must specify fund")
        url = self.base_url["fundfactsheet"].copy().add(path=[fund_id, "fund_compare"]).url
        result = self.__get_api_data(url)
        return result

    def get_class_fund(self, fund_id):
        if not fund_id:
            raise ValueError("Must specify fund")
        url = self.base_url["fundfactsheet"].copy().add(path=[fund_id, "class_fund"]).url
        result = self.__get_api_data(url)
        return result

    def get_fund_performance(self, fund_id):
        if not fund_id:
            raise ValueError("Must specify fund")
        url = self.base_url["fundfactsheet"].copy().add(path=[fund_id, "performance"]).url
        result = self.__get_api_data(url)
        return result

    def get_fund_5yearlost(self, fund_id):
        if not fund_id:
            raise ValueError("Must specify fund")
        url = self.base_url["fundfactsheet"].copy().add(path=[fund_id, "5YearLost"]).url
        result = self.__get_api_data(url)
        return result

    def get_fund_dividend_policy(self, fund_id):
        if not fund_id:
            raise ValueError("Must specify fund")
        url = self.base_url["fundfactsheet"].copy().add(path=[fund_id, "dividend"]).url
        result = self.__get_api_data(url)
        for record in result:
            if "dividend_details" in record:
                for dividend_record in record["dividend_details"]:
                    print(dividend_record)
                    print(type(dividend_record["book_closing_date"]))
                    if dividend_record["book_closing_date"] != "-":
                        dividend_record["book_closing_date"] = (
                            convert_buddhist_to_gregorian(dividend_record["book_closing_date"]).date().isoformat()
                        )
                    if dividend_record["payment_date"] != "-":
                        dividend_record["payment_date"] = (
                            convert_buddhist_to_gregorian(dividend_record["payment_date"]).date().isoformat()
                        )
        return result

    def get_fund_fee(self, fund_id):
        if not fund_id:
            raise ValueError("Must specify fund")
        url = self.base_url["fundfactsheet"].copy().add(path=[fund_id, "fee"]).url
        result = self.__get_api_data(url)
        return result

    def get_fund_involveparty(self, fund_id):
        if not fund_id:
            raise ValueError("Must specify fund")
        url = self.base_url["fundfactsheet"].copy().add(path=[fund_id, "InvolveParty"]).url
        result = self.__get_api_data(url)
        for record in result:
            if "effective_date" in record and record["effective_date"] != "-":
                record["effective_date"] = convert_buddhist_to_gregorian(record["effective_date"]).date().isoformat()
        return result

    def get_fund_port(self, fund_id, period):
        if not fund_id:
            raise ValueError("Must specify fund")
        url = self.base_url["fundfactsheet"].copy().add(path=[fund_id, "FundPort", period]).url
        result = self.__get_api_data(url)
        return result

    def get_fund_full_port(self, fund_id, period):
        if not fund_id:
            raise ValueError("Must specify fund")
        url = self.base_url["fundfactsheet"].copy().add(path=[fund_id, "FundPort", period]).url
        result = self.__get_api_data(url)
        return result

    def get_fund_top5_port(self, fund_id, period):
        if not fund_id:
            raise ValueError("Must specify fund")
        url = self.base_url["fundfactsheet"].copy().add(path=[fund_id, "FundTop5", period]).url
        result = self.__get_api_data(url)
        return result

    def get_fund_dividend(self, fund_id):
        if not fund_id:
            raise ValueError("Must specify fund")
        url = self.base_url["funddailyinfo"].copy().add(path=[fund_id, "dividend"]).url
        result = self.__get_api_data(url, subscription_key="funddailyinfo")
        return result

    def get_amc_submit_dailyinfo(self):
        url = self.base_url["funddailyinfo"].copy().add(path=["amc"]).url
        result = self.__get_api_data(url, subscription_key="funddailyinfo")
        return result
